patches/net/minecraft/client/particle/ParticleEngine.java.patch

```diff
@@ -1,14 +1,26 @@
 --- a/net/minecraft/client/particle/ParticleEngine.java
 +++ b/net/minecraft/client/particle/ParticleEngine.java
-@@ -28,7 +_,7 @@
-         ParticleRenderType.SINGLE_QUADS, ParticleRenderType.ITEM_PICKUP, ParticleRenderType.ELDER_GUARDIANS
+@@ -29,6 +_,8 @@
      );
      protected ClientLevel level;
--    private final Map<ParticleRenderType, ParticleGroup<?>> particles = Maps.newIdentityHashMap();
-+    private final Map<ParticleRenderType, ParticleGroup<?>> particles = Maps.newTreeMap(net.neoforged.neoforge.client.ClientHooks.makeParticleRenderTypeComparator(RENDER_ORDER));
+     private final Map<ParticleRenderType, ParticleGroup<?>> particles = Maps.newIdentityHashMap();
++    private final Map<ParticleRenderType, java.util.function.Function<ParticleEngine, ParticleGroup<?>>> particleGroupFactories;
++    private final List<ParticleRenderType> particleRenderOrder;
      private final Queue<TrackingEmitter> trackingEmitters = Queues.newArrayDeque();
      private final Queue<Particle> particlesToAdd = Queues.newArrayDeque();
      private final Object2IntOpenHashMap<ParticleLimit> trackedParticleCounts = new Object2IntOpenHashMap<>();
+@@ -38,6 +_,11 @@
+     public ParticleEngine(ClientLevel p_107299_, ParticleResources p_446012_) {
+         this.level = p_107299_;
+         this.resourceManager = p_446012_;
++        var particleGroupFactories = new java.util.HashMap<ParticleRenderType, java.util.function.Function<ParticleEngine, ParticleGroup<?>>>();
```

Comment:
I would prefer using an `IdentityHashMap` to replicate the vanilla behaviour of reference comparison and just remove the immutability. If you absolutely want the immutability, then you also have the option of using a `Reference2ObjectOpenHashMap` and wrapping it in `Reference2ObjectMaps.unmodifiable()`